### PR TITLE
K-nearest-neighbors

### PR DESCRIPTION
**Pull request by oliver-foggin**
Implemented `closest` using a k-nearest-neighbors algorithm.

This uses the structure of the quad tree to iteratively fetch points from child nodes. And skips any node that is further away than the furthest fetched point.

### CODE REVIEW

**Comment by oliver-foggin** [comment]
@crhallberg This is the k-nearest-neighbors algorithm I mentioned. With the fixed structure of the quad tree it can now be traversed recursively to fetch all the points within range.

File: test/rectangle.spec.js [hunk 1]
```diff
@@ -178,4 +178,26 @@ describe('Rectangle', () => {
       expect(rect.bottom).to.equal(210);
     });
   });
+  describe('distance from point', () => {
+    let rect;
+    before(() => {
+      rect = new Rectangle(5, 5, 10, 10);
+    });
+    it('calculates distance to point 1', () => {
```
Comment (crhallberg) [inline test/rectangle.spec.js:186]: Could we give these tests more meaningful names? "...to point inside" "...to point on edge", etc.

File: quadtree.js [hunk 2]
```diff
@@ -273,74 +322,31 @@ class QuadTree {
...
-    }
 
-    // Limit to number of points in this QuadTree
-    if (this.length == 0) {
-      return [];
-    }
-    if (this.length < count) {
-      return this.points;
-    }
+    var inRangePoints = [...this.points].filter((p) => p.distanceFrom(point) <= maxDistance);
```
Comment (crhallberg) [inline quadtree.js:330]: Is there an opportunity here to calculate `furthestDistance` inside of this filter?

File: quadtree.js [hunk 3]
```diff
@@ -273,74 +322,31 @@ class QuadTree {
...
+    inRangePoints.forEach(p => {
+        furthestDistance = Math.max(p.distanceFrom(point), furthestDistance);
+    });
+
+    furthestDistance = Math.min(furthestDistance, furthestFound);
+
+    this.children.forEach(child => {
+      if ((child.boundary.distanceFrom(point) < furthestFound || foundSoFar < count) &&
+          child.boundary.distanceFrom(point) < maxDistance) {
+        inRangePoints = inRangePoints.concat(child.closest(point, count, maxDistance, furthestDistance, inRangePoints.length + foundSoFar));
```
Comment (crhallberg) [inline quadtree.js:343]: Maybe I need to give a more careful reading but I'm unsure of the distinction between `furthestDistance` and `furthestFound`. If we don't want to check quads/children that are farther than the farthest found, should it be `Math.max` on line 338?

File: quadtree.js [hunk 4]
```diff
@@ -273,74 +322,31 @@ class QuadTree {
...
-        return aDist - bDist;
+    inRangePoints.forEach(p => {
+        furthestDistance = Math.max(p.distanceFrom(point), furthestDistance);
+    });
+
+    furthestDistance = Math.min(furthestDistance, furthestFound);
+
+    this.children.forEach(child => {
+      if ((child.boundary.distanceFrom(point) < furthestFound || foundSoFar < count) &&
+          child.boundary.distanceFrom(point) < maxDistance) {
```
Comment (crhallberg) [inline quadtree.js:342]: Style nitpick, how do you like the looks of this, which better matches the code styles elsewhere.

```JS
if (
  (blah || blah) &&
  blah
) {
  // code
}

**Review by crhallberg — CHANGES_REQUESTED** [review]
Just a few comments to try and clear things up. The algorithm seems to be in place, it's just a matter of making sure it's clear and easy to read.

File: quadtree.js (hunk 3 above)
Comment (oliverfoggin) [reply quadtree.js:343]: Yeah, I wasn't happy with the furthest found stuff and over night I fixed it in my head. Need to change it slightly. 😂

File: quadtree.js (hunk 2 above)
Comment (oliverfoggin) [reply quadtree.js:330]: Ah. Good idea. Yeah, I'll give it a go.

File: test/rectangle.spec.js (hunk 1 above)
Comment (oliver-foggin) [reply test/rectangle.spec.js:186]: Will do 👍🏻

**Comment by oliver-foggin** [comment]
Hey @crhallberg just checking if anything else needs to be updated with this? Thanks 👍🏻

**Comment by crhallberg** [comment]
This all looks very good, @oliver-foggin! Thank you! I'm gonna find time today to pull it down and give it a final look and test and then I'll merge it! Thank you!

**Comment by crhallberg** [comment]
Had the chance to review the code, at last! It looks good! Thanks again!

**Comment by crhallberg** [comment]
We may need to investigate this further. I started working on a new example to demonstrate how k-nearest works and I'm not getting encouraging results. Drop this into examples to see: [closest_vis_qtree.zip](https://github.com/CodingTrain/QuadTree/files/6159640/closest_vis_qtree.zip).

I think it's an issue of breadth-first vs. depth-first searching. Look at this one for example: https://ericandrewlewis.github.io/how-a-quadtree-works/.

**Comment by oliverfoggin** [comment]
Hmm, yeah, that's not ideal.

I'll take a look. It shouldn't be doing that. Once it has a point in the found points it should only search sub trees that are closer than it.

**Comment by oliver-foggin** [comment]
Hmm... @crhallberg I see in your code you have updated the tree.children.forEach and set all the points for all children to be true...

```
    tree.children.forEach((child) => {
        // if the child quad tree is further away from the point than the furthest point we have found so far
        // and we have enough points already
        // then we can skip checking this entire quad tree
        // as there will be no points in it that we are interested in.
        if (
            (child.boundary.distanceFrom(point) < furthestFound ||
                foundSoFar < count) &&
            child.boundary.distanceFrom(point) < maxDistance
        ) {
            inRangePoints = inRangePoints.concat(
                kNearest(
                    child,
                    point,
                    count,
                    maxDistance,
                    furthestFound,
                    inRangePoints.length + foundSoFar
                )
            );
            searchedQuads.push(child.boundary); // DISPLAY
        }

        child.points.forEach((p) => {
            p.userData.searched = true; // here
        }); // DISPLAY
    });
```

That's not what the function is doing as the children are only searched if they are closer than the furthest found point.

**Comment by oliver-foggin** [comment]
> I think it's an issue of breadth-first vs. depth-first searching. Look at this one for example: https://ericandrewlewis.github.io/how-a-quadtree-works/. It was up to me to catch this before I merged so I'm here to help with the solution of course.

This is the example I'm using. And the algorithm as it stands is (I believe) a depth first search.

I'll continue to have a look.

**Comment by crhallberg** [comment]
> Hmm... @crhallberg I see in your code you have updated the tree.children.forEach and set all the points for all children to be true...

Thanks! I fixed that display typo but it still shows what I feared. The purple quads were correct.

**Comment by oliver-foggin** [comment]
> > Hmm... @crhallberg I see in your code you have updated the tree.children.forEach and set all the points for all children to be true...
> 
> Thanks! I fixed that display typo but it still shows what I feared. The purple quads were correct.

Yeah, just looking into it now. One of the differences with the example in the link is that it is a balanced tree. The points only exist in the very bottom level of the branches. So it doesn't have to search points from higher up the branches. This makes the search easier to implement and faster to run.

Just seeing if there's a way around this now. I think the logic filtering the points is messed up. Looking into it now.

**Comment by crhallberg** [comment]
I got it working perfectly in my example function now but it's probably way over-engineered.

[closest_vis_qtree.zip](https://github.com/CodingTrain/QuadTree/files/6159974/closest_vis_qtree.zip)

**Comment by oliver-foggin** [comment]
> I got it working perfectly in my example function now but it's probably way over-engineered.
> 
> [closest_vis_qtree.zip](https://github.com/CodingTrain/QuadTree/files/6159974/closest_vis_qtree.zip)

You can't use the logic that the tree contains the point to filter out the children.

**Comment by oliver-foggin** [comment]
@crhallberg This works...

```
function kNearest(
    tree,
    searchPoint,
    maxCount = 1,
    maxDistance = Infinity,
    furthestDistance = 0,
    foundSoFar = 0
) {
    if (typeof searchPoint === "undefined") {
        throw TypeError("Method 'closest' needs a point");
    }

    let found = [];

    tree.points.forEach((p) => {
        const distance = p.distanceFrom(searchPoint);
        p.userData.searched = true;
        if (distance > maxDistance) {
            return;
        } else if (foundSoFar < maxCount || distance < furthestDistance) {
            found.push(p);
            furthestDistance = max(distance, furthestDistance);
            foundSoFar++;
        }
    });

    tree.children.forEach((child) => {
        const distance = child.boundary.distanceFrom(searchPoint);
        if (distance > maxDistance) {
            return;
        } else if (foundSoFar < maxCount || distance < furthestDistance) {
            const childPoints = kNearest(
                child,
                searchPoint,
                maxCount,
                maxDistance,
                furthestDistance,
                foundSoFar
            );
            searchedQuads.push(child.boundary);
            found = found.concat(childPoints);
        }
    });

    return found.sort((a, b) => a.distanceFrom(searchPoint) - b.distanceFrom(searchPoint))
        .slice(0, maxCount);
}
```

However, it still passes over more quads than it should. But that is down to the way that the tree is built.

If you have a tree with capacity 4 and put a point in each corner of it. Then wherever you press you will find those four points first. And so now you have to search the sub trees as they will all contain points closer than the original four. And so on down the structure to the bottom.

Having a balanced tree where all nodes are on the leaves will eliminate this issue.

**Comment by oliver-foggin** [comment]
You can see here that it is definitely removing subtrees from its search. The problem is that it can't limit the search enough due to the points in the parent trees.

image

I believe the code here is doing pretty much the same as the original code in the PR.

**Comment by oliver-foggin** [comment]
Actually... with a count of 1 the new code is actually much better...

**Comment by oliver-foggin** [comment]
Creating a new PR now.

**Pull request merged**